docs: Parallel Scraping guide (#2219)
apify/crawlee: docs/guides/parallel-scraping/parallel-scraping.mdx

---
id: parallel-scraping-guide
title: Parallel Scraping Guide
sidebar_label: Parallel Scraping
description: Parallelizing your scrapers with Crawlee
---

import ApiLink from '@site/src/components/ApiLink';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';

import SharedSource from '!!raw-loader!./shared.mjs';
import AdaptedRoutesSource from '!!raw-loader!./adapted-routes.mjs';
import ParallelScraperSource from '!!raw-loader!./parallel-scraper.mjs';
import ModifiedDetailRouteSource from '!!raw-loader!./modified-detail-route.mjs';

:::warning Experimental features ahead

> At the time of writing this guide (November 2023), request locking is still an experimental feature. You can read more about the experiment by visiting the [request locking experiment](../experiments/experiments-request-locking) page.

Reviewer: Shouldn't this say Decemeber 2023?

Author: Well... when I first wrote this it was still November 🤣, but time flew and I guess a december refresh wouldn't hurt

:::

In this guide, we will walk through how you can turn your single scraper into a scraper that can be parallelized and run in multiple instances. This guide
assumes you've read and walked through our [introduction guide](../introduction/setting-up) (or have a fully-fledged scraper already built), but if you haven't done so yet,
take a break, go read through all that, and come back. We'll be waiting...

*Oh, you're back already! Let's proceed in making that scraper parallel!*

## Things to consider before parallelizing

Before you rush ahead and change your scraper to support parallelization, take a minute to consider the following factors:

> - do you plan on scraping so many pages that you need to parallelize your scraper?

Reviewer: Is there any preference for use of - instead of * for unordered lists? I haven't seen it documented before so I added in style guide to use * and have been following this rule with all docs created since, but that can be easily changed and fixed

Reviewer 2: I think we use `-` usually, at least that's what I do as it's just one key away, unlike `*` which requires `shift` too :] So if we are picking one, I'd prefer `-` over `*` personally.

Let's not care about this now, we should add some code formatting that will fix this automatically. Let's first do it in the apify docs, then we can port it here.

Reviewer: I'll remove then the mention for this from style guide for now

```javascript title="src/routes.mjs"
import { getOrInitQueue } from './shared.mjs';
```

Then, replace the `CATEGORY` handler with the following:

<CodeBlock language="js" title="src/routes.mjs">{AdaptedRoutesSource}</CodeBlock>

Now, let's rename the file to `src/initial-scraper.mjs` and run it. You should see the crawler not scrape any detail pages, but now the URLs are being enqueued to the queue that supports locking!

Before we wrap up, let's also add the following line before `crawler.run()`:

```javascript title="src/initial-scraper.mjs"
import { getOrInitQueue } from './shared.mjs';

// Pre-initialize the queue so that we have a blank slate that will get filled out by the crawler
await getOrInitQueue(true);
```

We need this to ensure the queue always starts on an empty slate when we run the scraper. But you may not need this in your use case - remember to always experiment and see what works best!

And that's it with preparing our initial scraper to save all URLs we want to scrape to the queue that supports locking!

### Creating the parallel scrapers

> Up next, let's build out another scraper that will schedule the URLs from the queue to be scraped in parallel! For this, we will be using child processes from Node.js,

Reviewer 3: ```suggestion
Up next, let's build another scraper that will schedule the URLs from the queue to be scraped in parallel! For this, we will be using child processes from Node.js,
```

const config = new Configuration({
    storageClientOptions: {
        localDataDirectory: `./storage/worker-${process.env.WORKER_INDEX}`,
    },
});
```

#### Enabling the request locking experiment, and telling the crawler to use the worker configuration

You might have noticed several lines highlighted in the code above. Those show how you can enable the request locking experiment, as well as how you provide the request queue to the crawler.
You can read more about the experiment by visiting the [request locking experiment](../experiments/experiments-request-locking) page.

You might have also noticed we passed in a second parameter to the constructor of the crawler, the `config` variable we created earlier.
This is needed to ensure the crawler uses the worker-specific storages for internal states, and that they do not collide with each other.

#### Why do we use `process.send` instead of `Dataset.pushData`?

Since we use child processes, and each worker process has its own storage space, calling `Dataset.pushData` will not work as we want it to work (each worker would just

> push to it's own personal dataset that is considered the "default" one). Instead, we need to send the data back to the parent process, which has the dataset where we want to

Reviewer: ```suggestion
push to its own personal dataset that is considered the "default" one). Instead, we need to send the data back to the parent process, which has the dataset where we want to
```

store the data, in a centralized place.

:::tip Why don't we apply the same logic we did to the request queue to the dataset?

This is a very valid question, but it has a simple answer: since each process tracks its own internal state of how a dataset looks like (when we are scrapping locally),
the worker processes would get out of sync real fast and would either miss or override data. This is why we need to send the data back to the parent process, which has the dataset
where we want to store the data, in a centralized place.

Depending on your crawler, this might not be an issue! Each use case has its own quirks, but this is something you should keep in mind when building your scraper.

> :::

Reviewer 4: Hmm, this sounds like an "us" issue 😬😄 Given your expertise @vladfrangu , how much time would it take to make local Dataset parallel-friendly (I'm assuming the Apify Platform dataset isn't like this...?) I guess it would hurt the cache performance, right?

Author: Without fs storage of the dataset metadata, it's most likely impossible (maybe there's a way if we detect we can send data across threads but I do not want to cross that bridge of uncertainty and break who knows what). And given that storing said metadata is *optional* as well..

Realistically, this is an "issue" only with storages that rely on their process, and that aren't centralized. Apify wouldn't have this issue, a DB _probably_ wouldn't have this issue. To be fair, you'd need to store the requests on fs in the first place to have this work, so maybe we..actually enable storing metadata by default and then rely on that?

Also yes, it would possibly tank performance (and I am SCARED of the race conditions that that can cause)

Reviewer 4: Gotcha, no biggie then 👍🏽

#### Why did we limit the maximum concurrency to `5`?

This question has a two-fold answer:

- we don't want to overload the source website with requests, so we limit the number of concurrent requests to a reasonable number per worker process
- we don't want to overload the machine that is running the scraper

> This circles back to the initial paragraph about if you should parallelize your scraper or not. Remember: scrape responsibly!

Reviewer: ```suggestion
This circles back to the initial paragraph about whether you should parallelize your scraper or not. Remember: scrape responsibly!
```

## Other questions

> #### Couldn't the `initial-scraper` be merged into the `parallel-scraper`?

Reviewer: Wouldn't it be better to use _ emphasis for questions instead of h4?  Also if I remember correctly it's markdownlint violation to skip heading levels.

Reviewer 2: > Also if I remember correctly it's markdownlint violation to skip heading levels.

I guess we don't use that rule, as the linter is passing. Do you think it's really needed to enforce this? I think there are other places where it's done this way, one reason is that you don't want those things in the current page navigation (the floating one on the right side). Using emphasis would indeed fix it, but it might not be enough of a visual distinction. I'd maybe prefer block quotes.

Technically, it could! Nothing stops you from first enqueuing all the URLs in the parent process, and then run the worker process logic after to scrape them.
We separated them so it's easier to follow and understand what each part does, but you can merge them if you want to.

#### Will I benefit from this if I run XYZ scraper / want to scrape XYZ website?

We don't know! 🤷

> What we do know is that first, you should build your scraper to work as a single scraper, then monitor it's performance. Do you see it being too slow?

Reviewer 3: ```suggestion
What we do know is that first, you should build your scraper to work as a single scraper, then monitor its performance. Do you see it being too slow?
```

---
id: parallel-scraping-guide
title: Parallel Scraping Guide
sidebar_label: Parallel Scraping
description: Parallelizing your scrapers with Crawlee
---

import ApiLink from '@site/src/components/ApiLink';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';

import SharedSource from '!!raw-loader!./shared.mjs';
import AdaptedRoutesSource from '!!raw-loader!./adapted-routes.mjs';
import ParallelScraperSource from '!!raw-loader!./parallel-scraper.mjs';
import ModifiedDetailRouteSource from '!!raw-loader!./modified-detail-route.mjs';

:::warning Experimental features ahead

At the time of writing this guide (December 2023), request locking is still an experimental feature. You can read more about the experiment by visiting the [request locking experiment](../experiments/experiments-request-locking) page.

:::

> In this guide, we will walk through how you can turn your single scraper into a scraper that can be parallelized and run in multiple instances. This guide

Reviewer 5: It might be a personal preference, but I like to understand the problem before diving into a full article on a how-to. This helps determine if the subject might be related to what I'm looking for, especially when skimming through the docs.

In that sense, I suggest starting by introducing the problem, something like:
> Node.js is single-threaded for asynchronous I/O operations, but CPU-intensive tasks (like scraping with crawlee) can block the event loop, leading to performance issues. In such cases, running multiple Node.js scripts (or processes) can help. This approach aligns with the concept of horizontal scaling.

Then stating the "drawbacks" of the solution:

> However, there are considerations:

> - Complexity: Managing multiple processes adds complexity in terms of deployment, process management, inter-process communication, and error handling. For production workloads, you might want to use more robust tools like PM2 or a container orchestrator such as Kubernetes.
> - Resource Utilization: More processes mean more usage of system resources. You need to ensure that your system has enough memory and CPU resources.
> - Consistency and State Management: Since crawlers are running independently in their own process, they need to communicate to distribute the workload (the crawling requests) and thus require a shared state. Request locking has been introduced to solve this but it's still experimental.
> - Avoiding Overloading the Source Website: When parallelizing a web scraper, it's crucial to ensure that the scraping activity does not overload the source website. Running multiple scrapers in parallel can significantly increase the number of requests to the website, which might not only affect the website's performance but also lead to your IP being blocked.

Formatted this way, I know what this is about, what it solves, and what are the cons. If I'm still interested after weighing the pros and cons, *then* I proceed to read the how-to.

assumes you've read and walked through our [introduction guide](../introduction/setting-up) (or have a fully-fledged scraper already built), but if you haven't done so yet,

> take a break, go read through all that, and come back. We'll be waiting...

Author: Yes, it should be `we will walk you through`

As for prose wrap...I did them so we don't have 1 massive line for a paragraph (requiring the use of soft wrap in our editors). I'll revert them for now, but we have other guides with this in place 👀

*Oh, you're back already! Let's proceed in making that scraper parallel!*

## Things to consider before parallelizing

Before you rush ahead and change your scraper to support parallelization, take a minute to consider the following factors:

- do you plan on scraping so many pages that you need to parallelize your scraper?

>   > for example, if your scraper goes across few pages, you probably don't need parallelization  

Reviewer: ```suggestion
  > for example, if your scraper goes across a few pages, you probably don't need parallelization  
```

  > but if you scrape a lot of pages, or you scrape pages that take a long time to load, you might want to consider parallelization

> - can you parallelize your scraper while not overloading the source website?
>   > for example, if you scrape a website that has a lot of traffic, you don't want to add to that traffic by running multiple scrapers in parallel
>   > as that might cause the website to go down for all its users
> - do you have the resources available to run multiple scrapers in parallel?
>   > when running locally, depending on your scraper type, do you have enough CPU and RAM available to sustain multiple scrapers running in parallel  
>   > when running in the cloud, will the extra speed from parallelization be worth the extra cost of running multiple scrapers in parallel?

Reviewer: ```suggestion
- can you parallelize your scraper while not overloading the source website?
  - for example, if you scrape a website that has a lot of traffic, you don't want to add to that traffic by running multiple scrapers in parallel
  - as that might cause the website to go down for all its users
- do you have the resources available to run multiple scrapers in parallel?
  - when running locally, depending on your scraper type, do you have enough CPU and RAM available to sustain multiple scrapers running in parallel  
  - when running in the cloud, will the extra speed from parallelization be worth the extra cost of running multiple scrapers in parallel?
```

This way it would still be visually distinct,but it might look better? Especially the last question when you have 2 answers. They will be in separate bullet points instead of merged into one blockquote.

Author: Before and after

![Google Chrome - 2023-12-29 at 15 31 03@2x](https://github.com/apify/crawlee/assets/17960496/f219143f-27f7-4de6-a9b8-729169956f42)
![Google Chrome - 2023-12-29 at 15 31 46@2x](https://github.com/apify/crawlee/assets/17960496/ce2dc595-834d-409e-b026-79358ca19ceb)

>   > when running in the cloud, will the extra speed from parallelization be worth the extra cost of running multiple scrapers in parallel?

Author: Oof, the diff doesn't show it but the first line has 2 extra whitespaces at the end which cause it to insert a new line, but I see what you mean by the confusing aspects of this. Will fix

Let's assume you answered yes to all of those. Yes? Yes. Before we go ahead and get to the actual guide, we'd like to ask you to also take a read on [Apify's Ethical Web Scraping](https://blog.apify.com/what-is-ethical-web-scraping-and-how-do-you-do-it/)
blog post!

Now that we've gone through all that, the guide is split into two parts: converting the initial scraper we built in the [introduction guide](../introduction/setting-up) to one that prepares requests
to be usable in parallel scrapers, and then running scrapers in parallel.

:::note Want to see the final result?

You can see it on the [Crawlee Parallel Scraping Example](https://github.com/apify/crawlee-parallel-scraping-example) repository! It's the same scraper we built in the [introduction guide](../introduction/setting-up), but in TypeScript and
parallelized!

:::

### But isn't Crawlee already concurrent? What's the difference between concurrency and parallelization?

> > *Hold on! I've used Crawlee before, and it has a `maxConcurrency` option on it! What's this for then?!*

Reviewer 5: If using schemas can be considered, I think it's easier for the reader to visually understand the concepts.

> This what crawlee is doing by defaut (concurrency):
> ![image](https://github.com/apify/crawlee/assets/1524501/6008a2a7-3eed-444a-9d05-d95876fb10d0)
> This what we're achieving in this guide (parallelism):
> ![image](https://github.com/apify/crawlee/assets/1524501/6fcd534e-cefe-4c9c-8a25-29f3904907e1)

> > *Hold on! I've used Crawlee before, and it has a `maxConcurrency` option on it! What's this for then?!*

Reviewer: ```suggestion
> *Hold on! I've used Crawlee before, and it has a `maxConcurrency` option! What's this for then?!*
```

I think this sounds more natural

> > *Hold on! I've used Crawlee before, and it has a `maxConcurrency` option on it! What's this for then?!*

Reviewer: ```suggestion
> _Hold on! I've used Crawlee before, and it has a `maxConcurrency` option on it! What's this for then?!_
```

You're correct, Crawlee already supports scraping in "parallel" (more accurately called concurrent). What that enables is one process having multiple tasks that run in the background at the same time.

> But, as your scraping scales up in size, you will end up hitting bottlenecks, be them from the runtime simply not being able to process more requests at once, to resources like RAM and CPU being maxed out. You can only

Reviewer: ```suggestion
But, as your scraping operation scales up, you are likely to encounter bottlenecks. These can range from the runtime environment's inability to process more requests simultaneously, to resources like RAM and CPU being maxed out.
```

scale up resources so much before it stops providing a real benefit.

This is what people refer to when saying vertical or horizontal scaling. Vertical scaling is when you increase the resources of a single process or machine, while horizontal scaling is when you increase the number of processes or machines.

> And horizontal scaling is the kind of scaling (or what we're referring to as "parallelization") we are showcasing in this guide!

Reviewer: ```suggestion
Horizontal scaling, on the other hand, is the kind of scaling (or what we're referring to as "parallelization") we are showcasing in this guide!
```

## Preparing your scraper for parallelization

One of the best parts of Crawlee is that, for the most part, we do not need to change much to make this happen! Just create the queue that supports locking, enqueue links to it from the initial scraper,
then build scrapers that run in parallel that use that queue!

### Creating the request queue with locking support

> The first step in our conversion process will be creating a common file (let's call it `requestQueue.mjs`) that will store the request queue that supports request locking.

Author: I'm not a fan of using just `js`, it depends too much on what `type` is set to in the `package.json`, and can cause issues of "oh this project is CJS and now it's using ESM syntax in a js file but node is throwing" 😢. Also, I need to double check but i've definitely seen `mjs` in other guides too

<CodeBlock language="js" title="src/requestQueue.mjs">{SharedSource}</CodeBlock>

The exported function, `getOrInitQueue`, might seem like it does a lot. In essence, it just ensures the request queue is initialized, and if requested, ensures it starts off with an empty state.

### Adapting our previous scraper to enqueue the product URLs to the new queue

In the `src/routes.mjs` file of the scraper we previously built, we have a handler for the `CATEGORY` label. Let's adapt that handler to enqueue the product URLs to the new queue we created.

Firstly, let's import the `getOrInitQueue` function from the `requestQueue.mjs` file we created earlier. Add the following line at the start of the file:

> ```javascript title="src/routes.mjs"

Reviewer: ```suggestion
```javascript showLineNumbers title="src/routes.mjs"
```

I think this is good practice and nice QoL for people reading the docs/guides to see line numbers inside codeblocks. They should be added to every codeblock.

Before we wrap up, let's also add the following line before `crawler.run()`:

```javascript title="src/initial-scraper.mjs"
import { getOrInitQueue } from './requestQueue.mjs';

// Pre-initialize the queue so that we have a blank slate that will get filled out by the crawler
await getOrInitQueue(true);
```

We need this to ensure the queue always starts on an empty slate when we run the scraper. But you may not need this in your use case - remember to always experiment and see what works best!

And that's it with preparing our initial scraper to save all URLs we want to scrape to the queue that supports locking!

### Creating the parallel scrapers

Up next, let's build another scraper that will schedule the URLs from the queue to be scraped in parallel! For this, we will be using child processes from Node.js,
but you can use any other method you want to run multiple scrapers in parallel. You will need to adjust your code if you use other methods.

The scraper will fork itself twice (but you can experiment with this), and each fork will re-use the queue we created earlier.

> The best part? We can re-use the previous router we built for the initial scraper! Yay for code re-use!

Reviewer: ```suggestion
The best part? We can re-use the previous router we built for the initial scraper! Yay for code reuse!
```

// Get the request queue from the parent process (step 2)
const requestQueue = await getOrInitQueue(false);

// Configure crawlee to store the worker-specific data in a separate directory (needs to be done AFTER the queue is initialized when running locally) (step 3)
const config = new Configuration({
    storageClientOptions: {
        localDataDirectory: `./storage/worker-${process.env.WORKER_INDEX}`,
    },
});
```

#### Enabling the request locking experiment, and telling the crawler to use the worker configuration

You might have noticed several lines highlighted in the code above. Those show how you can enable the request locking experiment, as well as how you provide the request queue to the crawler.
You can read more about the experiment by visiting the [request locking experiment](../experiments/experiments-request-locking) page.

You might have also noticed we passed in a second parameter to the constructor of the crawler, the `config` variable we created earlier.
This is needed to ensure the crawler uses the worker-specific storages for internal states, and that they do not collide with each other.

#### Why do we use `process.send` instead of `context.pushData`?

> Since we use child processes, and each worker process has its own storage space, calling `context.pushData` will not work as we want it to work (each worker would just

Author: There's no way for us to remove this restriction unless we take control of how the child processes spawn (see the thread with @barjin https://github.com/apify/crawlee/pull/2219#discussion_r1417398874)

Its only an issue for this example, people running on, say, Apify, or that have a centralized API for it won't have this issue

push to its own personal dataset that is considered the "default" one). Instead, we need to send the data back to the parent process, which has the dataset where we want to
store the data, in a centralized place.

> :::tip Why don't we apply the same logic we did to the request queue to the dataset?

Reviewer: is tip the best type of admonition to use here? note or info seem more appropriate

Author: I'll go with info

This is a very valid question, but it has a simple answer: since each process tracks its own internal state of how a dataset looks like (when we are scrapping locally),
the worker processes would get out of sync real fast and would either miss or override data. This is why we need to send the data back to the parent process, which has the dataset
where we want to store the data, in a centralized place.

Depending on your crawler, this might not be an issue! Each use case has its own quirks, but this is something you should keep in mind when building your scraper.

:::

#### Why did we limit the maximum concurrency to `5`?

This question has a two-fold answer:

- we don't want to overload the source website with requests, so we limit the number of concurrent requests to a reasonable number per worker process
- we don't want to overload the machine that is running the scraper

> This circles back to the initial paragraph about whether you should parallelize your scraper or not. Remember: scrape responsibly!

Reviewer 2: the "Scrape responsibly" part is there again a few lines after this one, maybe we can remove it from here as it feels good in the very beginning of the guide. 

(not a blocker, just a thought)

:::

In this guide, we will walk you through how you can turn your single scraper into a scraper that can be parallelized and run in multiple instances. This guide assumes you've read and walked through our [introduction guide](../introduction/setting-up) (or have a fully-fledged scraper already built), but if you haven't done so yet, take a break, go read through all that, and come back. We'll be waiting...

*Oh, you're back already! Let's proceed in making that scraper parallel!*

## Things to consider before parallelizing

Before you rush ahead and change your scraper to support parallelization, take a minute to consider the following factors:

- Do you plan on scraping so many pages that you need to parallelize your scraper?

>   > For example, if your scraper goes across a few pages, you probably don't need parallelization<br/>

Reviewer 2: i still dont understand why you want the line break here, imo there should be none, if you want to have two paragraphs, you need an extra blank line instead of the `<br>` tag. maybe it should be a nested bullet list instead of quotes?

Author: Mostly for the way it looks (see https://github.com/apify/crawlee/pull/2219#discussion_r1438256142 - response to @TC-MO's review above yours).

Jus to get the ball rolling, I'll make it a nested list, it's a niche thing I'd rather not let hold back this PR

Reviewer 2: If it looks better, feel free to keep it, it just feels weird
